src/vmm/src/acpi/x86_64.rs

```diff
@@ -0,0 +1,53 @@
+// Copyright 2024 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::mem::size_of;
+
+use acpi_tables::fadt::{
+    IAPC_BOOT_ARG_FLAGS_MSI_NOT_PRESENT, IAPC_BOOT_ARG_FLAGS_PCI_ASPM,
+    IAPC_BOOT_ARG_FLAGS_VGA_NOT_PRESENT,
+};
+use acpi_tables::madt::{IoAPIC, LocalAPIC};
+use acpi_tables::{Aml, Fadt};
+use vm_memory::GuestAddress;
+use zerocopy::AsBytes;
+
+use crate::arch::x86_64::layout;
+use crate::Vmm;
+
+impl Vmm {
+    pub(crate) fn setup_interrupt_controllers(&self, nr_vcpus: u8) -> Vec<u8> {
+        let mut ic =
+            Vec::with_capacity(size_of::<IoAPIC>() + (nr_vcpus as usize) * size_of::<LocalAPIC>());
+
+        ic.extend_from_slice(IoAPIC::new(0, layout::IOAPIC_ADDR).as_bytes());
+        for i in 0..nr_vcpus {
+            ic.extend_from_slice(LocalAPIC::new(i).as_bytes());
+        }
+        ic
+    }
+
+    pub(crate) fn setup_arch_fadt(&self, fadt: &mut Fadt) {
+        // Let the guest kernel know that there is not VGA hardware present
+        // neither do we support ASPM, or MSI type of interrupts.
+        // More info here:
+        // https://uefi.org/specs/ACPI/6.5/05_ACPI_Software_Programming_Model.html?highlight=0a06#ia-pc-boot-architecture-flags
+        fadt.setup_iapc_flags(
+            1 << IAPC_BOOT_ARG_FLAGS_VGA_NOT_PRESENT
+                | 1 << IAPC_BOOT_ARG_FLAGS_PCI_ASPM
+                | 1 << IAPC_BOOT_ARG_FLAGS_MSI_NOT_PRESENT,
+        );
+    }
```

Comment:
So, the first one doesn't use `self` in x86, but it will in aarch64 because IRQ handler info is stored inside `Vmm` (in the `vm: Vm` field. I'd like to have the same structure in both architectures. I don't think there's any drawback in terms of performance: these things will most likely be inlined by the compiler (we can always add `#[inline(always)]` to be sure of it). The `setup_arch_fadt` indeed won't touch `self` but it just feels awkwrd to add some of the in the `impl Vmm` block and one not.

So, if we are to keep this code organization structure, I'd prefer keeping them like that.

Reply:
Yeah, I can't really think of functional reasons here, so happy to defer to your preference :)

Reply:
In v2 these are not part of an `impl Vmm` anymore.